Warn when eval or namespace calls are rendered, not when they are parsed

SunsetFi opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [x] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

<!--
If this PR resolves any issues, list them as



where 1234 is the issue number. This will help us with house-keeping as Github will automatically add a note to those issues stating that a potential fix exists. Once the PR is merged, Github will automatically close those issues.

Starting each line with a dash "-" will cause GitHub to display the issue title inline.

If an issue is only solved partially or is relevant in some other way, just list the number without "resolves".
-->

-  resolves #5881
 
### Description

Only emit warnings about calling eval() and namespaces if that call survives past tree-shaking.

To reduce duplication, each AST node will only warn once if it is rendered multiple times.

<!--
  Please be thorough and clearly explain the problem being solved.
  * If this PR adds a feature, look for previous discussion on the feature by searching the issues first.
  * Is this PR related to an issue?
-->

SunsetFi commented:
Is there documentation on contributing?  I couldn't find any, and I seem to be missing a lot of native tools required for both testing, and for local development.  As such, at the moment I am unable to run the tests.  Waiting to see what the github CI outputs...

wmertens commented:
@SunsetFi I found https://github.com/rollup/rollup/blob/master/CONTRIBUTING.md

SunsetFi commented:
Huh... I'm not sure how I missed that, given I was looking for just that filename...

At any rate, I got the tests working, and an interesting side effect of this is that the warnings are now arriving in reverse line order:

```
[test:only]        [
[test:only]       +  "main.js (4:1): Cannot call a namespace (\"foo\")."
[test:only]          "main.js (8:1): Cannot call a namespace (\"foo\")."
[test:only]       -  "main.js (4:1): Cannot call a namespace (\"foo\")."
[test:only]        ]
```

(Happens for both calling a namespace and for eval.)

I assume we want to preserve the line order for the warnings, but I'm not too sure on the timing of the render calls so I don't quite now how to approach a fix.

SunsetFi commented:
Scratch that, that was an easy fix.  I thought the test was two evals in the same scope but from the behavior I'm guessing its a scope difference.  Either way, just throwing after rendering the callee and arguments fixed it.

The only remaining error now is that onwarn is no longer being called... I'm not sure on the details of when or how this is called, but apparently the same log call on ast node bind isn't invoked during the render phase.

Is this expected?  The [docs for onwarn](https://rollupjs.org/configuration-options/#onwarn) just say it should get warnings, it doesn't mention anything about bind vs render phases.

TrickyPi commented:
Thanks! Currently we need to make the tests pass again. There are two points we should address:

1. The code in https://github.com/rollup/rollup/blob/master/src/ast/nodes/TaggedTemplateExpression.ts#L27-L34 should also be changed to fix the incorrect warning order.

2. Regarding the onwarn test: since the call to rollup.rollup doesn't include the render phase, we can call bundle.generate after that, like this:
```js
const bundle = await rollup.rollup({
	input: 'x',
	plugins: [loader({ x: `eval(42);` })],
	onwarn(warning, onwarn) {
		parameters = [warning, onwarn];
	}
});
await bundle.generate({ format: 'es' });
```

TrickyPi commented on src/ast/nodes/CallExpression.ts:
```diff
@@ -33,6 +33,9 @@ export default class CallExpression
 	/** Marked with #__PURE__ annotation */
 	declare annotationPure?: boolean;
 
+	private _evalWarned = false;
+	private _namespaceWarned = false;
```
To reduce memory overhead, we can convert this boolean flag to a bit flag, like the optional property below.

lukastaegert replied:
Following my other suggestion below, this could be

```suggestion
	private get hasCheckedForWarnings(): boolean {
		return isFlagSet(this.flags, Flag.checkedForWarnings);
	}

	private set hasCheckedForWarnings(value: boolean) {
		this.flags = setFlag(this.flags, Flag.checkedForWarnings, value);
	}
```

Then you just need to extend `BitFlags.ts` accordingly.

lukastaegert commented:
The tests are red because the order of warnings has changed in some places. Hopefully, this can be easily adapted in the tests.

SunsetFi commented:
Thanks for the feedback!  I implemented the memory saving, and fixed the remaining tests.

I also added an unused eval to warn-eval-multiple, but given the nature of that error i'm not entirely confident it is sufficient.  I think it would be good to create a brand new test explicitly for not warning on removed evals.  That would be testing for a lack of something, but when paired with the test that prove we get warnings for used evals,  I think it should be okay.

lukastaegert reviewed (approved):
Looks good to me, thanks a lot!
